### PR TITLE
Enhancements to the resource cache

### PR DESCRIPTION
**Pull request by kiiadi**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Description

Several enhancements to the resource cache:
* `Resource.Cache` keys now expose a `String` `id` field which is used as the cache-key - previously was using object identity.
* When a credential provider is updated or removed, its related cache entries are evicted
* Added two styles of blocking API: 
  * `getResourceNow` - a wrapper around the async API that will block until the resource is available
  * `getResourceIfPresent` - will return a value only if it exists in the cache (else `null`)

## Testing

new unit tests added to cover functionality

## Checklist

- [x] I have read the **README** document
- [x] I have read the **CONTRIBUTING** document
- [x] Local run of `gradlew check` succeeds
- [x] My code follows the code style of this project
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [ ] A short description of the change has been added to the **CHANGELOG**

## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: jetbrains-core/src/software/aws/toolkits/jetbrains/core/AwsResourceCache.kt [hunk 1]
```diff
@@ -50,6 +55,45 @@ interface AwsResourceCache {
...
+        credentialProvider: ToolkitCredentialsProvider,
+        timeout: Duration = DEFAULT_TIMEOUT,
+        useStale: Boolean = true,
+        forceFetch: Boolean = false
+    ): T = wait(timeout) { getResource(resource, region, credentialProvider, useStale, forceFetch) }
+
+    /**
+     * Gets the [resource] if it exists in the cache.
+     *
+     * @param[useStale] if an exception occurs attempting to refresh the resource return a cached version if it exists (even if it's expired). Default: true
```
Comment (abrooksv) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/core/AwsResourceCache.kt]: Statement about the exception doesnt seem to match impl

File: jetbrains-core/src/software/aws/toolkits/jetbrains/core/AwsResourceCache.kt [hunk 2]
```diff
@@ -159,28 +235,56 @@ class DefaultAwsResourceCache(private val project: Project, private val clock: C
...
-        } catch (e: Exception) {
-            LOG.warn(e) { "Failed to fetch resource using $key, falling back to expired entry" }
-            currentEntry
-        } as Entry<T>
+    override fun providerRemoved(providerId: String) = clearByCredential(providerId)
+
+    override fun providerModified(provider: ToolkitCredentialsProvider) = clearByCredential(provider.id)
+
+    private fun clearByCredential(providerId: String) {
+        cache.keys.filter { it.credentialsId == providerId }.forEach { cache.remove(it) }
```
Comment (abrooksv) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/core/AwsResourceCache.kt]: `.keys.removeIf` to simplify it

File: jetbrains-core/tst/software/aws/toolkits/jetbrains/core/AwsResourceCacheTest.kt [hunk 3]
```diff
@@ -212,6 +223,123 @@ class AwsResourceCacheTest {
...
+
+    @Test
+    fun canConditionallyFetchOnlyIfAvailableInCacheAndRespectExpiry() {
+        whenever(mockResource.fetch(any(), any(), any())).thenReturn("hello")
+
+        val now = Instant.now()
+        whenever(mockClock.instant()).thenReturn(now)
+        sut.getResource(mockResource, US_WEST_1, CRED1).value
+
+        whenever(mockClock.instant()).thenReturn(now.plus(DEFAULT_EXPIRY).plusMillis(1))
```
Comment (abrooksv) [inline jetbrains-core/tst/software/aws/toolkits/jetbrains/core/AwsResourceCacheTest.kt]: Subtract 1 milli to grantee it will always be in the past

Comment (kiiadi) [reply jetbrains-core/tst/software/aws/toolkits/jetbrains/core/AwsResourceCacheTest.kt]: this is current time - I want to make sure the expiry is always in the past hence current time is the expiry + 1 ms....

File: jetbrains-core/src/software/aws/toolkits/jetbrains/core/AwsResourceCache.kt (hunk 2 above)
Comment (kiiadi) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/core/AwsResourceCache.kt]: nice.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/core/AwsResourceCache.kt (hunk 1 above)
Comment (kiiadi) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/core/AwsResourceCache.kt]: will fix

File: jetbrains-core/tst/software/aws/toolkits/jetbrains/core/AwsResourceCacheTest.kt (hunk 3 above)
Comment (abrooksv) [reply jetbrains-core/tst/software/aws/toolkits/jetbrains/core/AwsResourceCacheTest.kt]: Ah, k

**Pull request merged**